otter/convergence/steps.py

```diff
@@ -539,3 +552,91 @@ def as_effect(self):
         """
         result = StepResult.LIMITED_RETRY if self.limited else StepResult.RETRY
         return Effect(Constant((result, list(self.reasons))))
+
+
+# ----- Cloud Orchestration Steps -----
+
+
+@implementer(IStep)
+@attr.s
+class CreateStack(object):
+    """
+    A stack must be created.
+
+    :ivar pmap stack_config: Heat launch configuration.
+
+    """
+    stack_config = attr.ib(validator=instance_of(PMap))
+
+    def as_effect(self, append_stack_uuid=append_stack_uuid):
```

Comment:
Okay, yeah, so this uuid4 call makes the function impure, which we don't want to do when it's a function that returns an Effect. (If it returns an Effect, the way it constructs that effect should be pure. The effect itself should represent all the side-effects to be performed). See `CreateServer.as_effect` for a very similar case.

So here's what I suggest:

    eff = Effect(Func(uuid4))
    def got_uid(uid):
        stack_config = append_stack_uuid(self.stack_config, uid)
        return create_stack(thaw(stack_config)).on(
            _success_reporter('Waiting for stack to create'))
    return eff.on(got_uid)

Or, if you want to avoid the explicit callback, you can use @do notation.